flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/SchemaResolutionTest.java

```diff
@@ -0,0 +1,366 @@
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.api.Schema;
+import org.apache.flink.table.api.TableConfig;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.expressions.ResolvedExpression;
+import org.apache.flink.table.expressions.resolver.ExpressionResolver;
+import org.apache.flink.table.expressions.utils.ResolvedExpressionMock;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.TimestampKind;
+import org.apache.flink.table.types.logical.TimestampType;
+import org.apache.flink.table.types.utils.DataTypeFactoryMock;
+import org.apache.flink.table.utils.FunctionLookupMock;
+
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Optional;
+
+import static org.apache.flink.table.api.Expressions.callSql;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.isProctimeAttribute;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.isRowtimeAttribute;
+import static org.apache.flink.table.types.utils.TypeConversions.fromLogicalToDataType;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/** Tests for {@link Schema}, {@link DefaultSchemaResolver}, and {@link ResolvedSchema}. */
+public class SchemaResolutionTest {
+
+    private static final String COMPUTED_COLUMN_SQL = "orig_ts - INTERVAL '60' MINUTE";
+
+    private static final ResolvedExpression COMPUTED_COLUMN_RESOLVED =
+            new ResolvedExpressionMock(DataTypes.TIMESTAMP(3), () -> COMPUTED_COLUMN_SQL);
+
+    private static final String WATERMARK_SQL = "ts - INTERVAL '5' SECOND";
+
+    private static final ResolvedExpression WATERMARK_RESOLVED =
+            new ResolvedExpressionMock(DataTypes.TIMESTAMP(3), () -> WATERMARK_SQL);
+
+    private static final String PROCTIME_SQL = "PROCTIME()";
+
+    private static final ResolvedExpression PROCTIME_RESOLVED =
+            new ResolvedExpressionMock(
+                    fromLogicalToDataType(new TimestampType(false, TimestampKind.PROCTIME, 3)),
+                    () -> PROCTIME_SQL);
+
+    private static final Schema SCHEMA =
+            Schema.newBuilder()
+                    .primaryKeyNamed("primary_constraint", "id") // out of order
+                    .column("id", DataTypes.INT().notNull())
+                    .column("counter", DataTypes.INT().notNull())
+                    .column("payload", "ROW<name STRING, age INT, flag BOOLEAN>")
+                    .columnByMetadata("topic", DataTypes.STRING(), true)
+                    .columnByMetadata("orig_ts", DataTypes.TIMESTAMP(3), "timestamp")
+                    .columnByExpression("ts", callSql(COMPUTED_COLUMN_SQL)) // API expression
```

Comment:
Could you put the computed column `ts` before `orig_ts`? I checked the implementation of `DefaultSchemaResolver#resolveColumns` and thought it doesn't take this case into account.

Reply:
I wasn't aware that we support this. Then also the website docs are wrong. I will fix them.